File: cs50x/week_06/credit.py
from cs50 import get_int


def main():
    # prompt the user for card number
    nm = get_int("Card number: ")

    # check for length
    i = len(str(nm))

    # check for legit card length and starting digits
    if i != 13 and i != 15 and i != 16:
        print("INVALID")
        return 0

    # checksum:
    # storing the sums for addition
    sum1 = 0
    sum2 = 0
    # dummy cc number for extraction
    chk = int(nm)

    # loop - extracting the digits and adding to the sum
    while True:
        # last digit
        ld = chk % 10
        # adding to sum1
        sum1 += ld
        # second to last digit
        chk = int(chk / 10)
        stld = chk % 10
        # multiply second to last digit by 2
        stld *= 2
        # extracting digits from the previous one multiplied by 2
        stld1 = int(stld / 10)
        stld2 = stld % 10
        # adding to sum2
        sum2 = sum2 + stld1 + stld2
        # to the next numbers
        chk = int(chk / 10)
        if chk <= 0:
            break

    # both sum1 and sum2
    sums = int(sum1 + sum2)

    # checking for the last digit of the total sums to be zero
    if sums % 10 != 0:
        print("INVALID")
        return 0

    # extracting first two digits of cc number
    ftdc = nm

    while True:
        ftdc = int(ftdc / 10)
        if ftdc < 100:
            break

    # checking for card type from the first two digits
    if (int(ftdc / 10) == 3) and (ftdc % 10 == 4 or ftdc % 10 == 7):
        print("AMEX")
    elif (int(ftdc / 10) == 5) and (ftdc % 10 < 6 and ftdc % 10 > 0):
        print("MASTERCARD")
    elif int(ftdc / 10) == 4:
        print("VISA")
    else:
        print("INVALID")
        return 0


main()

File: cs50x/week_01/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // prompt the user for card number
    long nm = get_long("Card number: ");

    // check for length
    int i = 0;
    long nmc = nm;
    while (nmc > 0)
    {
        nmc = nmc / 10;
        i++;
    }


    // check for legit card length and starting digits
    if (i != 13 && i != 15 && i != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // checksum:
    // storing the sums for addition
    int sum1 = 0;
    int sum2 = 0;
    // creditcard number copy for digit extraction
    long chk = nm;
    // last digit
    int ld;
    // second to last digit
    int stld;
    // and its two digits..
    int stld1;
    int stld2;
    // loop - extracting the digits and adding to the sum
    do
    {
        // last digit
        ld = chk % 10;
        // adding to sum1
        sum1 += ld;
        // second to last digit
        chk /= 10;
        stld = chk % 10;
        // multiply second to last digit by 2
        stld *= 2;
        // extracting digits from the previous one multiplied by 2
        stld1 = stld / 10;
        stld2 = stld % 10;
        // adding to sum2
        sum2 = sum2 + stld1 + stld2;
        // to the next numbers
        chk /= 10;
    }
    while (chk > 0);

    // both sum1 and sum2
    int sums = sum1 + sum2;

    // checking for the last digit of the total sums to be zero
    if (sums % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // extracting first two digits of cc number
    long ftdc = nm;
    do
    {
        ftdc /= 10;
    }
    while (ftdc > 100);

    // checking for card type from the first two digits
    if ((ftdc / 10 == 3) && (ftdc % 10 == 4 || ftdc % 10 == 7))
    {
        printf("AMEX\n");
    }
    else if ((ftdc / 10 == 5) && (ftdc % 10 < 6 && ftdc % 10 > 0))
    {
        printf("MASTERCARD\n");
    }
    else if (ftdc / 10 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}
